src/java_bytecode/java_static_initializers.cpp

```diff
@@ -135,82 +251,266 @@ static void create_clinit_wrapper_symbols(
     "clinit wrapper");
 }
 
-/// Produces the static initialiser wrapper body for the given function.
+/// Produces the static initialiser wrapper body for the given function. This
+/// static initializer implements (a simplification of) the algorithm defined
+/// in Section 5.5 of the JVM Specs. This function, or wrapper, checks whether
+/// static init has already taken place, calls the actual `<clinit>` method if
+/// not, and possibly recursively initializes super-classes and interfaces.
+/// Assume that C is the class to be initialized, then the algorithm is as
+/// follows:
+///
+/// \code
+///
+///   bool init_complete;
+///   if(java::C::__CPROVER_clinit_thread_local_state == INIT_COMPLETE)
+///   {
+///     return;
+///   }
+///   java::C::__CPROVER_clinit_thread_local_state = INIT_COMPLETE;
+///
+///   // This thread atomically checks and sets the global variable
+///   // 'clinit_state' in order to ensure that only this thread runs the
+///   // static initializer. The assume() statement below will prevent the SAT
+///   // solver from producing a thread schedule where more than 1 thread is
+///   // initializing the initializer. At the end of this function the only
+///   // thread that runs the static initializer will update the variable.
+///   // Alternatively we could have done a busy wait / spin-lock, but that
+///   // would achieve the same effect and blow up the size of the SAT formula.
+///   ATOMIC_BEGIN
+///   assume(java::C::__CPROVER_clinit_state != IN_PROGRESS)
+///   if(java::C::__CPROVER_clinit_state == NOT_INIT)
+///   {
+///     java::C::__CPROVER_clinit_state = IN_PROGRESS
+///     init_complete = false;
+///   }
+///   else if(java::C::__CPROVER_clinit_state == INIT_COMPLETE)
+///   {
+///     init_complete = true;
+///   }
+///   ATOMIC_END
+///
+///   if(init_complete)
+///     return;
+///
+///   java::C'::clinit_wrapper();
+///   java::I1::clinit_wrapper();
+///   java::I2::clinit_wrapper();
+///   // ...
+///   java::In::clinit_wrapper();
+///
+///   java::C::<clinit>();
+///
+///   // Setting this variable to INIT_COMPLETE will let other threads "cross"
+///   // beyond the assume() statement above in this function.
+///   ATOMIC_START
+///   C::__CPROVER_clinit_state = INIT_COMPLETE;
+///   ATOMIC_END
+///
+///   return;
+///
+///  \endcode
+///
+/// Note: The current implementation does not deal with exceptions.
```

Comment:
This seems fairly serious - is there a ticket to fix this (if so, could you provide a reference here)

Reply:
The previous version also didn't handle exceptions. There is a performance vs soundness  balancing act aspect here as you have to be careful not to blowup the SAT formula. Having said that I am all for creating an issue if you feel this needs to be investigated further.

Reply:
As your in the area I will trust your judgement on this matter :+1: